AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+                                                      const AudioTimeStamp *timestamp,
+                                                      UInt32 busNumber,
+                                                      UInt32 numFrames,
+                                                      AudioBufferList *bufferList) {
+    assert(bufferList->mNumberBuffers == 1);
+    assert(bufferList->mBuffers[0].mNumberChannels <= 2);
+    assert(bufferList->mBuffers[0].mNumberChannels > 0);
+
+    RendererAudioContext *context = (RendererAudioContext *)refCon;
+    context->bufferList = bufferList;
+
+    int8_t *audioBuffer = (int8_t *)bufferList->mBuffers[0].mData;
+    UInt32 audioBufferSizeInBytes = bufferList->mBuffers[0].mDataByteSize;
+
+    // Pull decoded, mixed audio data from the media engine into the AudioUnit's AudioBufferList.
+    assert(numFrames <= context->maxFramesPerBuffer);
+    assert(audioBufferSizeInBytes == (bufferList->mBuffers[0].mNumberChannels * kAudioSampleSize * numFrames));
+    OSStatus outputStatus = noErr;
+
+    if (@available(iOS 11.0, *)) {
+        AVAudioEngineManualRenderingBlock renderBlock = (__bridge AVAudioEngineManualRenderingBlock)(context->renderBlock);
+        const AVAudioEngineManualRenderingStatus status = renderBlock(numFrames, bufferList, &outputStatus);
+
+        /*
+         * Render silence if there are temporary mismatches between CoreAudio and our rendering format or AVAudioEngine
+         * could not render the audio samples.
+         */
+        if (numFrames > context->maxFramesPerBuffer ||
+            status != AVAudioEngineManualRenderingStatusSuccess) {
+            if (numFrames > context->maxFramesPerBuffer) {
+                NSLog(@"Can handle a max of %u frames but got %u.", (unsigned int)context->maxFramesPerBuffer, (unsigned int)numFrames);
+            }
+            *actionFlags |= kAudioUnitRenderAction_OutputIsSilence;
+            memset(audioBuffer, 0, audioBufferSizeInBytes);
+        }
+    }
+
+    return noErr;
+}
+
+static OSStatus ExampleCoreAudioDeviceCaptureCallback(void *refCon,
+                                                      AudioUnitRenderActionFlags *actionFlags,
+                                                      const AudioTimeStamp *timestamp,
+                                                      UInt32 busNumber,
+                                                      UInt32 numFrames,
+                                                      AudioBufferList *bufferList) {
+
+    if (numFrames > kMaximumFramesPerBuffer) {
+        NSLog(@"Expected %u frames but got %u.", (unsigned int)kMaximumFramesPerBuffer, (unsigned int)numFrames);
+        return noErr;
+    }
+
+    CapturerAudioContext *context = (CapturerAudioContext *)refCon;
+    AudioBufferList *audioBufferList = context->bufferList;
+    audioBufferList->mBuffers[0].mDataByteSize = numFrames * sizeof(UInt16) * kPreferredNumberOfChannels;
+    audioBufferList->mBuffers[0].mData = NULL;
```

Comment:
Okay, I guess I misunderstood. Is this causing the AudioUnit's preallocated buffer to be used in AudioUnitRender?

Reply:
Right AudioUnit fills the buffer for us. 

BTW I learned it from this example - https://github.com/GoogleCloudPlatform/ios-docs-samples/blob/cda5b8089fb741b3e54ef83dca9d4bfe7000c47a/speech/Objective-C/Speech-gRPC-Streaming/Speech/AudioController.m#L77